subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/ImplementationSnapshot.java

```diff
@@ -69,51 +83,274 @@ public ValueSnapshot snapshot(Object value, ValueSnapshotter snapshotter) {
+
+    private static class DefaultImplementationSnapshot extends ImplementationSnapshot {
+        private final HashCode classLoaderHash;
+
+        public DefaultImplementationSnapshot(String typeName, HashCode classLoaderHash) {
+            super(typeName);
+            this.classLoaderHash = classLoaderHash;
         }
-        if (o == null || getClass() != o.getClass()) {
-            return false;
+
+        @Override
+        public void appendToHasher(BuildCacheHasher hasher) {
+            hasher.putString(ImplementationSnapshot.class.getName());
+            hasher.putString(getTypeName());
+            hasher.putHash(classLoaderHash);
         }
 
-        ImplementationSnapshot that = (ImplementationSnapshot) o;
+        protected boolean isSameSnapshot(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
+            DefaultImplementationSnapshot that = (DefaultImplementationSnapshot) o;
+
+            if (!getTypeName().equals(that.getTypeName())) {
+                return false;
+            }
+            return classLoaderHash.equals(that.classLoaderHash);
+        }
 
-        if (!typeName.equals(that.typeName)) {
+        public HashCode getClassLoaderHash() {
+            return classLoaderHash;
+        }
+
+        public boolean isUnknown() {
             return false;
         }
-        return classLoaderHash != null ? classLoaderHash.equals(that.classLoaderHash) : that.classLoaderHash == null;
+
+        @Nullable
+        public String getUnknownReason() {
+            return null;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            DefaultImplementationSnapshot that = (DefaultImplementationSnapshot) o;
+            if (this == o) {
+                return true;
+            }
+
+
+            if (!getTypeName().equals(that.getTypeName())) {
+                return false;
+            }
+            return classLoaderHash.equals(that.classLoaderHash);
+        }
+
+        @Override
+        public int hashCode() {
+            int result = super.hashCode();
+            result = 31 * result + classLoaderHash.hashCode();
+            return result;
+        }
+
+        @Override
+        public String toString() {
+            return getTypeName() + "@" + classLoaderHash;
+        }
     }
 
-    @Override
-    public boolean equals(Object o) {
-        if (o == null || getClass() != o.getClass()) {
-            return false;
+    private static class LambdaImplementationSnapshot extends ImplementationSnapshot {
+
+        public LambdaImplementationSnapshot(String typeName, HashCode classLoaderHash) {
+            super(typeName);
+            this.classLoaderHash = classLoaderHash;
```

Comment:
Why do we need the classloader hash for a lambda?

Reply:
I was thinking maybe that gives some useful information when comparing inputs via build scans. But you are probably right, it just adds to the noise. I'll remove it.

Reply:
Removed.